File: backend/mallapi/src/main/java/org/nareun130/mallapi/service/ProductServiceImpl.java
package org.nareun130.mallapi.service;

import java.util.List;
import java.util.stream.Collectors;

import org.nareun130.mallapi.domain.Product;
import org.nareun130.mallapi.domain.ProductImage;
import org.nareun130.mallapi.dto.PageRequestDTO;
import org.nareun130.mallapi.dto.PageResponseDTO;
import org.nareun130.mallapi.dto.ProductDTO;
import org.nareun130.mallapi.repository.ProductRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import lombok.RequiredArgsConstructor;
import lombok.extern.log4j.Log4j2;

@Service
@Log4j2
@RequiredArgsConstructor
@Transactional
public class ProductServiceImpl implements ProductService{

    private final  ProductRepository productRepository;

    @Override
    public PageResponseDTO<ProductDTO> getList(PageRequestDTO pageRequestDTO) {
        
        log.info("getList ...................");

        Pageable pageable = PageRequest.of(
            pageRequestDTO.getPage() -1 , 
            pageRequestDTO.getSize(), 
            Sort.by("pno").descending());
       
        //* Object[] 내용 : Product객체와 ProductImage객체
        Page<Object[]> result = productRepository.selectList(pageable);

        List<ProductDTO> dtoList = result.get().map(arr -> {
            Product product = (Product)arr[0];
            ProductImage productImage = (ProductImage)arr[1];

            ProductDTO productDTO = ProductDTO.builder()
                .pno(product.getPno())
                .pname(product.getPname())
                .pdesc(product.getPdesc())
                .price(product.getPrice())
                .build();
            
            String imageStr = productImage.getFileName();
            productDTO.setUploadFileNames(List.of(imageStr));

            return productDTO;
        }).collect(Collectors.toList());

        long totalCount = result.getTotalElements();

        return PageResponseDTO.<ProductDTO>withAll()
            .dtoList(dtoList)
            .totalCount(totalCount)
            .pageRequestDTO(pageRequestDTO)
            .build();
    }

    @Override
    public Long register(ProductDTO productDTO) {
        
        Product product = dtoToEntity(productDTO);

        Product result = productRepository.save(product);

        return result.getPno();
        
    }

    private Product dtoToEntity(ProductDTO productDTO) {

        Product product = Product.builder()
            .pno(productDTO.getPno())
            .pname(productDTO.getPname())
            .pdesc(productDTO.getPdesc())
            .price(productDTO.getPrice())
            .build();

            //* 업로드 처리가 끝난 파일들의 이름 리스트
            List<String> uploadFileNames = productDTO.getUploadFileNames();

            if(uploadFileNames == null) {
                return product;
            }

            uploadFileNames.stream().forEach(uploadName -> {
                product.addImageString(uploadName);
            });

            return product;
    }
    
}

File: backend/mallapi-kotlin/src/main/kotlin/com/nareun/mallapikotlin/service/ProductServiceImpl.kt
package com.nareun.mallapikotlin.service

import com.nareun.mallapikotlin.domain.Product
import com.nareun.mallapikotlin.domain.ProductImage
import com.nareun.mallapikotlin.dto.PageRequestDTO
import com.nareun.mallapikotlin.dto.PageResponseDTO
import com.nareun.mallapikotlin.dto.ProductDTO
import com.nareun.mallapikotlin.repository.ProductRepository
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class ProductServiceImpl : ProductService {

    @Autowired
    private lateinit var productRepository: ProductRepository

    val log = LoggerFactory.getLogger(ProductServiceImpl::class.java)

    override fun getList(pageRequestDTO: PageRequestDTO): PageResponseDTO<ProductDTO> {
        log.info("getList..............")

        val pageable: Pageable = PageRequest.of(
            //페이지 시작 번호가 0 부터 시작하기 때문
            pageRequestDTO.page - 1,
            pageRequestDTO.size,
            Sort.by("pno").descending()
        )

        val result: Page<Array<Any>> = productRepository.selectList(pageable)

        val dtoList = result.map { arr ->
            val product = arr[0] as Product// 형변환
            val productImage = arr[1] as ProductImage


            ProductDTO(
                pno = product.pno,
                pname = product.pname,
                pdesc = product.pdesc,
                price = product.price,
                uploadFileNames = listOf(productImage.fileName)
            )
        }.content.toList()

        return PageResponseDTO.withAll(
            dtoList = dtoList,
            totalCount = result.totalElements,
            pageRequestDTO = pageRequestDTO
        )
    }

    override fun register(productDTO: ProductDTO): Long {
        val product : Product = dtoToEntity(productDTO)

        val result : Product = productRepository.save(product)

        return result.pno
    }

    private fun dtoToEntity(productDTO: ProductDTO): Product {
        val product = Product(
            pno = productDTO.pno,
            pname = productDTO.pname,
            pdesc = productDTO.pdesc,
            price = productDTO.price,
        )
        //업로드 처리가 끝난 파일들의 이름 리스트
        val uploadFileNames : List<String> ?= productDTO.uploadFileNames//? 처리 안할 시 오류

        uploadFileNames?.forEach { fileName -> product.addImageString(fileName) }

        return product
    }

    override fun get(pno: Long): ProductDTO {
        TODO("Not yet implemented")
    }

    override fun modify(productDTO: ProductDTO) {
        TODO("Not yet implemented")
    }

    override fun remove(pno: Long) {
        TODO("Not yet implemented")
    }

}